tactics/tactics.ml

```diff
@@ -1781,12 +1780,20 @@ let make_projection env sigma params cstr sign elim i n c u =
         && not (isEvar sigma (fst (whd_betaiota_stack env sigma t)))
         && (not (isRel sigma t))
       then
-        let t = lift (i+1-n) t in
-        let (elim, _) = eval_case_analysis elim in
-        let abselim = beta_applist sigma (elim, params@[t;branch]) in
-        let args = Context.Rel.instance mkRel 0 sign in
-        let c = beta_applist sigma (abselim, [mkApp (c, args)]) in
+        let (_, mip) = Inductive.lookup_mind_specif env ind in
+        let t = lift (i + 1 - n) t in
+        let ksort = Retyping.get_sort_family_of (push_rel_context sign env) sigma t in
+        if Sorts.family_leq ksort mip.mind_kelim then
+          let arity = List.firstn mip.mind_nrealdecls mip.mind_arity_ctxt in
+          let mknas ctx = Array.of_list (List.rev_map get_annot ctx) in
+          let ci = Inductiveops.make_case_info env ind (get_relevance decl) RegularStyle in
+          let br = [| mknas cs_args, b |] in
+          let args = Context.Rel.instance mkRel 0 sign in
+          let pnas = Array.append (mknas arity) [|make_annot Anonymous mip.mind_relevance|] in
+          let p = (pnas, lift (Array.length pnas) t) in
+          let c = mkCase (ci, u, Array.of_list params, p, NoInvert, mkApp (c, args), br) in
```

Comment:
Would calling `build_case_analysis_scheme` here, where we now have the right sort IIUC, make sense?

Reply:
This is generating a lot of cruft and creates dummy β-redexes even after head reduction, so not a good idea.

Reply:
Right, dummy β-redexes pas glop.

What worried me at first is that this is a piece of code that is morally semantically stand-alone and which is inlined in a way which makes the whole code more difficult to read.

Reply:
I guess it's a matter of taste, but the previous code also required a bit of juggling.

Reply:
> I guess it's a matter of taste, but the previous code also required a bit of juggling.

Possibly, but as it is, I find it more difficult to read in diagonal than before.